src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -322,6 +324,299 @@ public virtual void Encode(TextWriter output, char[] value, int startIndex, int
+            }
+
+            // Input buffer has been fully processed!
+
+            charsConsumed = originalSourceLength;
+            charsWritten = originalDestinationLength - destination.Length;
+            return OperationStatus.Done;
+
+        ReturnDestinationTooSmall:
+
+            charsConsumed = originalSourceLength - source.Length;
+            charsWritten = originalDestinationLength - destination.Length;
+            return OperationStatus.DestinationTooSmall;
+
+        ReturnNeedMoreData:
+
+            charsConsumed = originalSourceLength - source.Length;
+            charsWritten = originalDestinationLength - destination.Length;
+            return OperationStatus.NeedMoreData;
+        }
+
+        public unsafe virtual OperationStatus EncodeUtf8(ReadOnlySpan<byte> utf8Source, Span<byte> utf8Destination, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
+        {
+            // Optimization: Detect how much "doesn't require escaping" data exists at the beginning of the buffer,
+            // and memcpy it directly to the destination.
+
+            int numBytesToCopy = FindFirstCharacterToEncodeUtf8(utf8Source);
+            if (numBytesToCopy < 0)
+            {
+                numBytesToCopy = utf8Source.Length;
+            }
+
+            if (!utf8Source.Slice(0, numBytesToCopy).TryCopyTo(utf8Destination))
+            {
+                // There wasn't enough room in the destination to copy over the entire source buffer.
+                // We'll instead copy over as much as we can and return Incomplete. We do need to
+                // account for the fact that we don't want to truncate a multi-byte UTF-8 subsequence
+                // mid-sequence (since a subsequent slice and call to EncodeUtf8 would produce invalid
+                // data).
+
+                utf8Source = utf8Source.Slice(0, utf8Destination.Length + 1); // guaranteed not to fail since utf8Source is larger than utf8Destination
+                for (int i = utf8Source.Length - 1; i >= 0; i--)
+                {
+                    if (!UnicodeHelpers.IsUtf8ContinuationByte(in utf8Source[i]))
+                    {
+                        utf8Source.Slice(0, i).CopyTo(utf8Destination);
+                        bytesConsumed = i;
+                        bytesWritten = i;
+                        return OperationStatus.DestinationTooSmall;
+                    }
+                }
+
+                // If we got to this point, either somebody mutated the input buffer out from under us, or
+                // the FindFirstCharacterToEncodeUtf8 method was overridden incorrectly such that it attempted
+                // to skip over ill-formed data. In either case we don't know how to perform a partial memcpy
+                // so we shouldn't do anything at all.
+
+                bytesConsumed = 0;
+                bytesWritten = 0;
+                return OperationStatus.DestinationTooSmall;
```

Comment:
Is that the right `OperationStatus` for such cases? Should it instead of `InvalidData`, or maybe even an explicit throw?

Reply:
Making the destination buffer bigger (at least the size of the source buffer) would resolve the issue, so this seems like the appropriate status to return. Good eye, though! I'll add a comment clarifying this.